docs/proposals-accepted/20240909-live-blocks-in-objectstorage.md

```diff
@@ -0,0 +1,62 @@
+---
+type: proposal
+title: Avoid querying blocks in object storage if they have a live source
+status: approved
+owner: MichaHoffmann
+menu: proposals-accepted
+---
+
+## Avoid querying blocks in object storage if they have a live source
+
+> TL;DR: We propose to not query blocks in object storage that are still served by a Sidecar/Ruler or Receiver.
+
+## Why
+
+Accessing blocks from object storage is generally orders of magnitude slower then accessing the same data from other sources that can query them from disk or from memory. If the source that originally uploaded the blocks is still alive and still has access to the block then we would throw away the expensively obtained data from object storage during deduplication.
+
+## How
+
+### Solution
+
+Each source utilizes the `shipper` component to upload blocks into object storage. Additionally the `shipper` component also has a complete picture of the blocks that the source owns. If we extend the `shipper` to also own a register in object storage named `thanos/sources/<uuid>/live_blocks.json` that contains a plain list of the live blocks that this source owns. We can deduce a timestamp when it was last updated by checking attributes in object storage. When the storage gateway syncs its list of block metas, it can also iterate the `thanos/sources` directory and see which `live_blocks.json` files have been updated recently enough to assume that the sources are still alive. It can subsequently build an index of live block ids and proceed to prune them when it handles Store API requests (Series, LabelValues, LabelNames RPCs). In theory this should not lead to gaps since the block_ids are still owned by other live sources.
```

Comment:
Ok, so correct me if I'm wrong here, but the last update for the live_blocks.json file, would be the moment a new block is cut, which is maybe 2h durations.

I think this would be too long of a duration to assume your sidecar/ruler/receive is still up?

Reply:
We can write this file more often maybe?

Reply:
We could send the current live IDs with series API maybe, that way we dont need any heartbeats